opnode/l1/downloader.go

```diff
@@ -2,291 +2,69 @@ package l1
-		return
-	}
-	// stop fetching when the task is cancelled or when the individual receipt times out
-	ctx, cancel := context.WithTimeout(task.dest.ctx, fetchReceiptTimeout)
-	defer cancel()
-	receipt, err := dl.src.TransactionReceipt(ctx, task.txHash)
-	if err != nil {
-		// if a single receipt fails out of the whole block, we can retry a few times.
-		if task.retry >= maxReceiptRetry {
-			// Failed to get the receipt too many times, block fails!
-			task.dest.Finish(wrappedErr{fmt.Errorf("failed to download receipt again, and reached max %d retries: %v", maxReceiptRetry, err)})
-			return
-		} else {
-			task.retry += 1
-			select {
-			case dl.receiptTasks <- task:
-				// all good, retry scheduled successfully
-				return
-			default:
-				// failed to schedule, too much receipt work, stop block to relieve pressure.
-				task.dest.Finish(wrappedErr{fmt.Errorf("receipt downloader too busy, not downloading receipt again (%d retries): %v", task.retry, err)})
-				return
-			}
-		}
-	}
-	task.dest.receipts[task.txIndex] = receipt
-	// We count the receipts we have so far (atomic, avoid parallel counting race condition)
-	total := atomic.AddUint32(&task.dest.downloadedReceipts, 1)
-	if total == uint32(len(task.dest.receipts)) {
-		// block completed without error!
-		task.dest.Finish(wrappedErr{nil})
-		return
-	}
-	// task completed, but no Finish call without other receipt tasks finishing first
-}
-
-func (dl *downloader) newReceiptWorker() ethereum.Subscription {
-	return event.NewSubscription(func(quit <-chan struct{}) error {
-		for {
-			select {
-			case task := <-dl.receiptTasks:
-				dl.processTask(task)
-			case <-quit:
-				return nil
-			}
-		}
-	})
-}
-
-// AddReceiptWorkers can add or remove (negative value) worker routines to parallelize receipt downloads with.
-// It returns the number of active workers.
-func (dl *downloader) AddReceiptWorkers(n int) int {
-	dl.receiptWorkersLock.Lock()
-	defer dl.receiptWorkersLock.Unlock()
-	if n < 0 {
-		for i := 0; i < -n && len(dl.receiptWorkers) > 0; i++ {
-			last := len(dl.receiptWorkers) - 1
-			dl.receiptWorkers[last].Unsubscribe()
-			dl.receiptWorkers = dl.receiptWorkers[:last]
-		}
-	}
-	for i := 0; i < n; i++ {
-		dl.receiptWorkers = append(dl.receiptWorkers, dl.newReceiptWorker())
-	}
-	return len(dl.receiptWorkers)
-}
-
-// Close unsubscribes and removes the receipt workers, then drains all remaining tasks
-func (dl *downloader) Close() {
-	// nil the src to stop accepting new download tasks
-	dl.cacheLock.Lock()
-	if dl.src == nil {
-		return // already closed
-	}
-	dl.src = nil
-	dl.cacheLock.Unlock()
-
-	dl.receiptWorkersLock.Lock()
-	defer dl.receiptWorkersLock.Unlock()
-	for _, w := range dl.receiptWorkers {
-		w.Unsubscribe()
-	}
-	dl.receiptWorkers = dl.receiptWorkers[:0]
-	close(dl.receiptTasks)
-	for task := range dl.receiptTasks {
-		task.dest.Finish(wrappedErr{DownloadClosedErr})
-	}
+	wg.Wait()
+	return block, receipts, retErr
```

Comment:
To make reasoning about this function easier, I think we should return `nil, nil, retErr` if `retErr != nil`. This way either all of the receipt fetches succeed, or the function returns an error. I doubt callers will use the value of `receipts` when `retErr != nil` anyway since they could contain unexpected nil values.